core/src/main/java/org/mule/DefaultMuleMessage.java

```diff
@@ -760,7 +761,10 @@ public String getPayloadAsString(String encoding) throws Exception
         String result = getPayload(DataType.STRING_DATA_TYPE, encoding);
         if (muleContext.getConfiguration().isCacheMessageAsBytes())
         {
-            cache = result.getBytes(encoding);
+            if (isValidUT8Character(result))
```

Comment:
I suppoused the encoding condition is implicit (if the encoding isn't utf-8, the � character wouldn't exist). However, this condition could be added in the if statement explicitly.